src/components/implementation/no_interface/vkernel/vkernel.c

```diff
@@ -0,0 +1,195 @@
+#include <cos_component.h>
+#include <cobj_format.h>
+#include <cos_kernel_api.h>
+
+#undef assert
+#define assert(node) do { if (unlikely(!(node))) { debug_print("assert error in @ "); *((int *)0) = 0; } } while (0)
+#define PRINT_FN prints
+#define debug_print(str) (PRINT_FN(str __FILE__ ":" STR(__LINE__) ".\n"))
+#define BUG() do { debug_print("BUG @ "); *((int *)0) = 0; } while (0);
+
+struct vms_info {
+	struct cos_compinfo cinfo;
+	captblcap_t ct;
+	pgtblcap_t pt, utpt;
+	compcap_t cc;
+	thdcap_t initthd, exitthd;
+	thdid_t inittid;
+	tcap_t inittcap;
+	arcvcap_t initrcv;
+} vmx_info[COS_VIRT_MACH_COUNT];
+
+struct vkernel_info {
+	struct cos_compinfo cinfo;
+
+	thdcap_t termthd;
+	asndcap_t vminitasnd[COS_VIRT_MACH_COUNT];
+} vk_info;
+
+extern vaddr_t cos_upcall_entry;
+extern void vm_init(void *);
+
+unsigned int ready_vms = COS_VIRT_MACH_COUNT;
+struct cos_compinfo *vk_cinfo = (struct cos_compinfo *)&vk_info.cinfo;
+
+void
+vk_terminate(void *d)
+{
+	BUG();
+}
+
+void
+vm_exit(void *d)
+{
+	printc("%d: EXIT\n", (int)d);
+	ready_vms --;
+	vmx_info[(int)d].initthd = 0;	
+
+	cos_thd_switch(BOOT_CAPTBL_SELF_INITTHD_BASE);
+}
+
+void
+scheduler(void) 
+{
+	static unsigned int i;
+	thdid_t tid;
+	int rcving;
+	cycles_t cycles;
+	int index;
+
+	while (ready_vms) {
+		cos_sched_rcv(BOOT_CAPTBL_SELF_INITRCV_BASE, &tid, &rcving, &cycles);
+		index = i++ % COS_VIRT_MACH_COUNT;
+		
+		if (vmx_info[index].initthd && vk_info.vminitasnd[index]) {
```

Comment:
If a Vm at `index` has exited, only then I set initthd = 0. 
And what you mentioned can happen, this would wait for next timer event and wouldn't schedule any other VM. But my concern was, if I have nested for/while here to check for next VM, I might be spending a lot of time here to find a vm to activate. And the periodic timer interval is 100us / 1ms in rump2cos, so it went for simplicity. 

Actually VM is woken up by vkernel scheduler, that means, when it reports that it is rcving, vkernel scheduler is the one responsible to be sending to it (virt-timer event). All I could do was, block for this immediate event and then maybe send to it the next time I index this VM. But that would also mean, I'd have to have another nested loop around cos_sched_rcv to receive all events to get all such blocked events and then mark them blocked for this real timer event and then unblock them at the next real timer event I'd get. All this increasing the scheduler's complexity, so I didn't do it for the first version.